File: backend/src/groots/domain/model/album.py
from dataclasses import dataclass, field
from datetime import datetime
from enum import StrEnum

from groots.domain.model.base import new_id


class RecordingFormat(StrEnum):
    CD = "CD"
    LP = "LP"
    EP = "EP"
    SINGLE = "Single"
    COMPILATION = "Compilation"
    DIGITAL = "Digital"
    CASSETTE = "Cassette"


class Genre(StrEnum):
    BLUES = "Blues"
    BRASS_AND_MILITARY = "Brass & Military"
    CHILDRENS = "Children's"
    CLASSICAL = "Classical"
    ELECTRONIC = "Electronic"
    FOLK_WORLD_AND_COUNTRY = "Folk, World, & Country"
    FUNK_SOUL = "Funk / Soul"
    HIP_HOP = "Hip Hop"
    JAZZ = "Jazz"
    LATIN = "Latin"
    NON_MUSIC = "Non-Music"
    POP = "Pop"
    REGGAE = "Reggae"
    ROCK = "Rock"
    STAGE_AND_SCREEN = "Stage & Screen"


@dataclass
class Album:
    title: str
    artist: str
    id: str = field(default_factory=new_id)
    year: int | None = None
    genre: Genre | None = None
    description: str | None = None
    cover_cid: str | None = None  # IPFS CID for cover artwork
    recording_format: RecordingFormat | None = None  # e.g. CD, LP, EP, Single…
    user_id: str | None = None  # owner of this album
    created_at: datetime = field(default_factory=datetime.utcnow)

    def __post_init__(self) -> None:
        if self.recording_format is not None and not isinstance(
            self.recording_format, RecordingFormat
        ):
            try:
                self.recording_format = RecordingFormat(self.recording_format)
            except ValueError:
                self.recording_format = None
        if self.genre is not None and not isinstance(self.genre, Genre):
            try:
                self.genre = Genre(self.genre)
            except ValueError:
                self.genre = None

File: groots_app/lib/src/domain/models/album.dart
const recordingFormats = ['CD', 'LP', 'EP', 'Single', 'Compilation', 'Digital', 'Cassette'];

class Album {
  final String id;
  final String title;
  final String artist;
  final int? year;
  final String? genre;
  final String? description;
  final String? coverCid;
  final String? recordingFormat;
  final String? userId;

  const Album({
    required this.id,
    required this.title,
    required this.artist,
    this.year,
    this.genre,
    this.description,
    this.coverCid,
    this.recordingFormat,
    this.userId,
  });

  factory Album.fromJson(Map<String, dynamic> json) => Album(
        id: json['id'] as String,
        title: json['title'] as String,
        artist: json['artist'] as String,
        year: json['year'] as int?,
        genre: json['genre'] as String?,
        description: json['description'] as String?,
        coverCid: json['cover_cid'] as String?,
        recordingFormat: json['recording_format'] as String?,
        userId: json['user_id'] as String?,
      );
}
